### ISSUE TITLE
【推理】vLLM 启动时加载 LLM 全过程解析

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#202]
> 本文基于 vLLM v0.13.0 版本

# 整体流程图

<img width="1594" height="598" alt="Image" src="https://github.com/user-attachments/assets/1036776c-e244-46a0-8bd1-e3dade8e9ddf" />

# 具体过程

## 0. 入口：CLI 解析与参数归一化

https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/api_server.py#L113

- argparse 解析 CLI 参数
- 组装 EngineArgs（引擎参数）+ API server 参数
- 打印 non- default args

## 1. 构建引擎配置（EngineArgs --> vllmConfig）

vLLM 会把参数拆成多个子配置：[^1]
- ModelConfig：模型/Tokenizer 相关（模型路径、dtype、max_model_len、trust_remote_code、rope_scaling、quantization 等）
- ParallelConfig：TP/PP、worker 数、分布式后端
- DeviceConfig：cuda/rocm/cpu，device 映射
- CacheConfig：KV cache（块大小、swap space、cpu offload 等）
- SchedulerConfig：批处理/抢占/预填充策略（chunked prefill、max_num_batched_tokens 等）
- LoadConfig：权重加载方式（safetensors、分片、流式等）
- 以及 observability / structured output / speculative 等可选 config

这个阶段也会做一些“前置探测”：
- 解析/校验 max_model_len
- 依据模型 config 推导架构，如 Qwen3ForCausalLM （因果模型架构）
- 根据 CPU 内存计算 swap space

## 2. 启动 API Server（仅 vllm serve）

vllm serve 会启动一个 OpenAI 风格的 HTTP 服务（uvicorn/fastapi）：
- 创建监听端口
- 在后台准备引擎（异步 build engine client）
- 等引擎 ready 后开始接收请求

注意：真正加载模型通常发生在引擎进程/worker里，而不是主 server 进程里。

代码：https://github.com/vllm-project/vllm/blob/v0.13.0/vllm/entrypoints/openai/api_server.py#L1391

```
async def run_server(args, **uvicorn_kwargs) -> None:
    """Run a single-worker API server."""

    # Add process-specific prefix to stdout and stderr.
    decorate_logs("APIServer")

    listen_address, sock = setup_server(args)
    await run_server_worker(listen_address, sock, args, **uvicorn_kwargs)
```

## 3. 初始化分布式与 Worker

> Ray / multiprocessing / torch.distributed

如果设置了 `--tensor-parallel-size N`（tensor 并行，例如单机多卡），vLLM 会：
1. 选择执行模式（本机多进程为主；也可 Ray）
3. 拉起 N 个 worker（通常每个 worker 绑定一张卡）
4. 初始化 NCCL 通信
5. 给每个worker 分配 rank、local_rank、device id

这个过程中，Ray 负责控制面（调度/拉进程/绑GPU/跨节点），NCCL 管数据面（多卡之间怎么高速交换张量）。

这一步常见问题：
- NCCL 互联/IB 设置不当
- GPU 可见性、device ordinal 不一致
- 进程数/线程数限制（pids-limit、ulimit、OpenMP/OpenBLAS 线程）

**注意**：tensor-parallel 是拆每一层的矩阵，pipeline-parallel 是按层拆。

## 4. 加载 Tokenizer 与模型配置（config/tokenizer 的解析）

流程是：
1. 找模型目录/仓库：
    - 本地目录：直接读目录
    - HF repo id：走 huggingface_hub 下载到 cache，一般默认目录是：`~/.cache/huggingface/hub`
3. 读取：
    - config.json（架构、hidden size、rope、num_layers...）
    - tokenizer 相关（tokenizer.json / vocab.json+merge.txt / tokenizer_config.json 等）
5. 应用 --hf-overrides，可覆盖 config.json 中的配置
6. 确定 max_model_len 最终值

## 5. 构建模型计算图与权重容器

> PyTorch Module 初始化

worker 内大致做：
1. 根据 architecture（例如 Qwen3ForCausalLM）创建 nn.Module ，其中 vLLM 对某模型的适配工作也主要是这块在做。[^2]
2. 设置 dtype（fp16/bf16/fp32）
3. 初始化线性层/注意力层等结构（此时权重还可能是空壳或占位）

## 6. 选择/编译算子与内核

> Kernel / Backend 选择
> 这里 Kernel / backend 指的是：在同一个“高层算子”（例如注意力、GEMM/Linear、RMSNorm、采样等）下面，vLLM 可以选用不同的 底层实现路径。这些实现路径可能是：
> - PyTorch 自带的 CUDA kernel（cuBLAS / cuDNN / ATen）
> - 第三方高性能 kernel（FlashAttention、xFormers）
> - vLLM 自己的 CUDA/C++ 扩展（vllm._C）
> - Triton 写的 kernel（运行时 JIT 编译）
> - 针对量化格式的专用 kernel（AWQ marlin、GPTQ 等）
>
> 它们都是“最终在 GPU 上执行的具体实现”，只是来源/优化手段不同。vLLM 会在启动或首次用到时做选择。

vLLM 会根据 GPU 架构、是否量化（AWQ、FP8）、是否可用 Triton / FlashAttention / xFormers 来选择实现路径。部分内核需要 JIT 编译（Triton / torch.compile），首次启动会慢。

例如，如果模型是 awq 量化的，vLLM 可能会选择 awq_marlin kernel，因为 AWQ 权重走 marlin 内核更快。

## 7. 权重加载

> safetensors 分片 --> GPU/CPU [^3]

典型流程：
1. 读取权重索引：model.safetensors.index.json（如果分片）
2. 按层/按分片加载：
    - 先落 CPU 内存，再搬运到 GPU
    - TP 模式会把每层权重按 shard 切分到不同 rank（rank 是分布式并行语义）
3. 对量化权重做必要的 unpack/format 转换：awq 可能会把权重转换到适合 marlin 的布局。

这一步最容易遇到：
- GPU 显存不足
- CPU 内存不足/NUMA/IO 慢
- 权重文件损失/缺失
- 目录里多了奇怪同名文件，例如 `._model.safetensors.index.json`

## 8. 分配 KV Cache

> 显存规划的关键

vLLM 的吞吐核心在 KV cache 管理。启动时会：
1. 根据 gpu-memory-utilization 估算可以用于 KV 的显存
    - 算法：用“总显存 * 利用率” 作为 vLLM 可支配的显存上限，然后在这个上限内先放模型权重/运行时开销，剩下的尽量都用于 KV cache（block pool）。KV 能分配多少最终取决于“预算减去已占用”。
3. 为每张卡分配 KV blocks（paged attention）
4. 若启用 swap/offload，则在 CPU 内存预留 swap space

## 9. Scheduler 启动

> 请求调度与 chunked prefill

启动后 scheduler 会常驻，负责：
- 把多个请求合并 batch
- 预填充（prefill）与解码（decode）阶段的 token 调度
- chunked prefill ，用于长上下文场景减少峰值

## 10. 服务 Ready：对外提供 OpenAI API

当：
- workers 全部 ready
- 权重加载完成
- KV cache 分配完成

API server 才会进入可用状态，开始响应：
- /v1/chat/completions
- /v1/completions
- /v1/models 等

[^1]: https://docs.vllm.ai/en/v0.11.2/cli/serve/
[^2]: https://github.com/vllm-project/vllm/tree/v0.13.0/vllm/model_executor/models
[^3]: https://github.com/vllm-project/vllm/blob/v0.13.0/vllm/model_executor/model_loader/weight_utils.py
